accepted/00000-auto-generate-bootstrap-repositories.md

```diff
@@ -0,0 +1,91 @@
+
+A bootstrap repository provide the software for the supported operating systems with its dependencies. The content for the
+bootstrap repositories come from the normal channels which must be synchronized and updated on regular base.
+
+Users very often do not know this, or forget to create them when they synchronize a new products and channels.
+Also the bootstrap repository is only updated when the user execute the generator script again.
+
+This RFC should describe how to change the current tools to automatically 
+execute the generation of bootstrap repositories after a new product or channel was added,
+and after a channel changed because the underlying repositories synced new or updated packages.
+
+# Detailed design
+[design]: #detailed-design
+
+The currently exiting tool `mgr-create-bootstrap-repo` list all available bootstrap repositories it could generate.
+This list depends on the products/channels the user has synchronized.
+
+Step 1: provide an option to enable an automatic mode which iterate over all available bootstrap repositories and generate them
+when needed. As we have sometimes have multiple options for the same local "path" we should iterate over the paths to minimize
+the number of runs.
+
+Step 2: add a log file to `mgr-create-bootstrap-repo`. Currently it has no log file. But when this tool is running in the background
+we need the possibility to debug problems. All messages should go to a log file.
+
+Step 3: let `mgr-create-bootstrap-repo` use a configuration file. All commandline options should be possible to set via
+configuration file. This allows to influence the behavior even in automatic mode when it is not possible to change the commandline
+options.
+
+Step 4: implement a locking mechanism to run only 1 instance of `mgr-create-bootstrap-repo` at the same time or more fine grained
+to prevent generating the same repository from 2 instances. As triggering from reposync already serialize the generation of bootstrap
+repositories, we need to take care only about manual called `mgr-create-bootstrap-repo`. A full lock on the whole application
+should be sufficient here.
+
+To trigger the re-generation we implement the following algorithm:
+
+- call `mgr-create-bootstrap-repo` from a successful finished reposync
+- It calculate all channels which are needed to create the bootstrap repository for every available OS
+- It compare the `last_modified` timestamps of the channels (DB) with the `modified date` of the main metadata file
+  (repomd.xml or Release)
+- if all `last_modified` timestamps are newer then the timestamp of the main metadata file, re-generate the bootstrap repository
+- if any `last_modified` timestamps are newer then the timestamp of the main metadata file and the `modified date` of the main metadata
+  file is older than 26 hours, re-generate the bootstrap repository. This is for the case a channel is not synced. It can happen on
```

Comment:
Why this complex logic and why the 26 hours? Given that `mgr-create-bootstrap-repo` will be called from reposync, why not just "if any of the `last_modified` timestamps for one of the repos of that channel is newer that the last modified bootstrap repository?" I don't think we need to make a difference between custom repos and vendor repos.

Reply:
Only when all repos are synced you can get a successful run of the tool.

Example - first time sync after the product was added:

- tools update channel finished sync => error with not found packages
- tools pool channel finished sync => error with not found packages
- sles pool channel finished sync => success, but you may have outdated packages
- sles update channel finished sync => success, now you are really up to date.

With notifications it would produce at least 2 errors and user start L3s.

Also in case update just syncing updates there is the chance that new required packages cause errors.

The 26 hours come from the initial idea to run it at least 1 time per day. I think with the current approach we can go down to some hours (4) and check that all repos were synced at least 1 time before.

Reply:
Sorry but I do not understand it yet.

I get the "all repos were synced at least 1 time before", I'm in 100%.

But why the 4 hours? Where does that magic number come from?

You cannot count on a repository being synced only once a day. User may trigger reposync manually and if new/updated packages come, bootstrap repository needs to be created/updated.

Reply:
> But why the 4 hours? Where does that magic number come from?

When we talk about custom repositories the user is free to disable syncing. When he knows that the content does not change he could decide to configure it that way.
When this happens, we never reach all repos re-synced. So we need a number as a grace period we wait (maybe it is just syncing later) before we just re-generate it even if not all repos were re-synced.

If you think it is ok that we generate them even when not all repos are re-synced. We can do this, but it would cause 4 times re-generation of the repo for every normal SLE product.

With this approach I wanted to limit it and try to generate it only when we are really "complete" and have a consistant state.
Getting an inconsistant state will happen only 1-2 times per year. But we know it will happen.

Reply:
I am all for whatever avoids inconsistent states :-)